### DISCUSSION TITLE
`usePrepareContractWrite`+`write` with variable arguments

### DISCUSSION THREAD

**Discussion by vrde** [wevm/wagmi#2626] [Q&A]
I'm refactoring my code to use `usePrepareContractWrite` instead of `useContractRead`, but I'm having issues with the type of the `write` function returned by `useContractRead`, as it seems to not accept any parameter.

My code looks more or less like this:

```typescript
export default function Mint({ tokenId, ...props }: ButtonProps & { tokenId?: string }) {
  const { isOpen, onOpen, onClose } = useDisclosure();

  const { owner, isMine, priceUsd, priceEth, isLoaded } = useToken(tokenId);
  const { address } = useAccount();

  const { config } = usePrepareContractWrite({
    ...contractConfig,
    functionName: "mint",
    value: 0n,
  });

  const contract = useContractWrite(config);

  const transaction = useWaitForTransaction({
    hash: contract.data?.hash,
  });


  function mint() {
    onOpen();
    if (!address || !tokenId || !contract.write) {
      return;
    }

    // The problem is here 🔥
    contract.write({
      args: [address, BigInt(tokenId)],
      value: priceEth?.toBigInt(),
    });
  }
```

But I get a type error

<img width="455" alt="image" src="https://github.com/wagmi-dev/wagmi/assets/134680/8384cf66-1580-4e48-9c8e-a976f6d55972">

Note: by using only `useContractWrite`, typescript is able to infer the type of the `mint` function correctly.

**Comment by frank-coder** [comment]
Did you find a solution to this ?

**Reply by vrde** [reply]
Unfortunately not :(

Will try to update wagmi later and see if I still have the issue. If so, I'll open a bug.

**Reply by gope-gope** [reply]
Experiencing the same. Not sure how to pass parameters to a function.

**Reply by dataexcess** [reply]
I have the same.... it's difficult to abstract away usePreparecontractWrite/write in a separate file / method and pass arguments T_T

**Reply by xavierdmello** [reply]
Same issue here

**Reply by Ethaan** [reply]
So i was having this same mental problem, think about `usePrepare..` as a hook that prepares the write to happen, in your case, the data is `static` (or non reactive), so the prepare should look like

```
 const { config } = usePrepareContractWrite({
    ...contractConfig,
    functionName: "mint",
    args: [address, BigInt(tokenId)],
    value: priceEth?.toBigInt(),
  });
```

and then you can even use `writeSync` to do `await contract.writeSync`

if you need args of value to change depending on, lets say a `useState` value, i would recommend doing a `debounce` for `usePrepareContractWrite` args param
